src/core/data/transform.cc

```diff
@@ -185,6 +187,13 @@ void Promote::print(std::ostream& out) const
 
 int32_t Promote::target_ndim(int32_t source_ndim) const { return source_ndim - 1; }
 
+void Promote::return_promoted_dims(std::vector<int32_t>& dims) const
+{
+  // avoid pushing 2 times
+  auto finder = std::find(dims.begin(), dims.end(), (extra_dim_));
+  if (finder == dims.end()) { dims.push_back(extra_dim_); }
```

Comment:
```suggestion
  for (auto& dim : dims) if (dim >= extra_dim_) dim++;
  dims.push_back(extra_dim_);
```

IIUC the added dimension is "bumping up" all the existing dimensions. Consider this example:

```
x = cn.ones((4,5)) # shape=(4,5), promoted_dims=[]
x = x[:,np.newaxis,:] # shape=(4,1,5), promoted_dims=[1]
x = x[np.newaxis,:,:,:] # shape=(1,4,1,5), promoted_dims=[0,2]
```

Therefore, I believe all existing promoted dims should remain, and adjusted up if they came after the added dim.